[ONNX][dynamo_export] Decomposition skips using custom operator

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117314

A context manager that disables the decomposition of certain ops during dynamo tracing.

The approach is to temporarily hijack the operator callable with PT2 custom operator.
The custom operator will not be decomposed and will show up as a single node to be exported to ONNX.

For the time being the decomposition of these ops is otherwise unavoidable.

https://github.com/pytorch/pytorch/issues/116684
https://github.com/pytorch/pytorch/issues/115883

This solution will no longer be required once the issue is resolved.

BowenBao committed:
[ONNX][dynamo_export] Decomposition skips using custom operator

[ghstack-poisoned]

BowenBao committed:
lint that was somehow missed on "[ONNX][dynamo_export] Decomposition skips using custom operator"


A context manager that disables the decomposition of certain ops during dynamo tracing.

The approach is to temporarily hijack the operator callable with PT2 custom operator.
The custom operator will not be decomposed and will show up as a single node to be exported to ONNX.

For the time being the decomposition of these ops is otherwise unavoidable.

https://github.com/pytorch/pytorch/issues/116684
https://github.com/pytorch/pytorch/issues/115883

This solution will no longer be required once the issue is resolved.

[ghstack-poisoned]

BowenBao committed:
Fix typo on "[ONNX][dynamo_export] Decomposition skips using custom operator"


A context manager that disables the decomposition of certain ops during dynamo tracing.

The approach is to temporarily hijack the operator callable with PT2 custom operator.
The custom operator will not be decomposed and will show up as a single node to be exported to ONNX.

For the time being the decomposition of these ops is otherwise unavoidable.

https://github.com/pytorch/pytorch/issues/116684
https://github.com/pytorch/pytorch/issues/115883

This solution will no longer be required once the issue is resolved.

[ghstack-poisoned]

thiagocrepaldi reviewed (commented):
Nice! Would it make sense to have this exact API, but exposed to the user instead of having it implicitly used by the exporter? e.g. the torchbench would be changed to add this around the torch.onnx.dynamo_export call and leave the exporter as-is, without temporary stuff

```python
with torch.onnx._enabled_decomposition_skips():
   onnx_program = torch.onnx.dynamo_export()
```

The advantage is that we don't need another PR just to disable this when the bug is fixed. We can also reuse this private API for other operators as they show up without changing the exporter too
Maybe we could even tap this solution along with the custom op/decomposition table features too

BowenBao committed:
Fix tests, lint and onnxscript version on "[ONNX][dynamo_export] Decomposition skips using custom operator"


A context manager that disables the decomposition of certain ops during dynamo tracing.

The approach is to temporarily hijack the operator callable with PT2 custom operator.
The custom operator will not be decomposed and will show up as a single node to be exported to ONNX.

For the time being the decomposition of these ops is otherwise unavoidable.

https://github.com/pytorch/pytorch/issues/116684
https://github.com/pytorch/pytorch/issues/115883

This solution will no longer be required once the issue is resolved.

[ghstack-poisoned]

BowenBao committed:
Create torch.library.Library object to tie lifetime to context on "[ONNX][dynamo_export] Decomposition skips using custom operator"


A context manager that disables the decomposition of certain ops during dynamo tracing.

The approach is to temporarily hijack the operator callable with PT2 custom operator.
The custom operator will not be decomposed and will show up as a single node to be exported to ONNX.

For the time being the decomposition of these ops is otherwise unavoidable.

https://github.com/pytorch/pytorch/issues/116684
https://github.com/pytorch/pytorch/issues/115883

This solution will no longer be required once the issue is resolved.

[ghstack-poisoned]

BowenBao committed:
backout from lib: non-deterministic errors including segfault and ValueError: vector reserve on "[ONNX][dynamo_export] Decomposition skips using custom operator"


A context manager that disables the decomposition of certain ops during dynamo tracing.

The approach is to temporarily hijack the operator callable with PT2 custom operator.
The custom operator will not be decomposed and will show up as a single node to be exported to ONNX.

For the time being the decomposition of these ops is otherwise unavoidable.

https://github.com/pytorch/pytorch/issues/116684
https://github.com/pytorch/pytorch/issues/115883

This solution will no longer be required once the issue is resolved.

[ghstack-poisoned]

albanD commented on test/test_public_bindings.py:
```diff
@@ -250,6 +250,7 @@ def test_modules_can_be_imported(self):
             "torch._inductor.codegen.cuda.cuda_kernel",
             "torch.onnx._internal.fx._pass",
             "torch.onnx._internal.fx.analysis",
+            "torch.onnx._internal.fx.decomposition_skip",
```
It would be best to make this module not crash on import.

BowenBao replied:
@albanD could you approve? Please let me know if this is a concern.

BowenBao committed:
Bump onnxscript on "[ONNX][dynamo_export] Decomposition skips using custom operator"


A context manager that disables the decomposition of certain ops during dynamo tracing.

The approach is to temporarily hijack the operator callable with PT2 custom operator.
The custom operator will not be decomposed and will show up as a single node to be exported to ONNX.

For the time being the decomposition of these ops is otherwise unavoidable.

https://github.com/pytorch/pytorch/issues/116684
https://github.com/pytorch/pytorch/issues/115883

This solution will no longer be required once the issue is resolved.

[ghstack-poisoned]

justinchuby commented on torch/onnx/_internal/fx/decomposition_skip.py:
```diff
@@ -0,0 +1,139 @@
+"""The namespace for the custom operator."""
+
+
+class DecompSkip:
```
```suggestion
class DecompSkip(abc.ABC):
```

was this abc?

BowenBao replied:
I'll likely add this in a follow up now that all tests are green.. Depending on if we have permission to merge.

justinchuby reviewed (approved):
This is neat!

thiagocrepaldi commented:
> Hmm I actually would prefer adding a PR to disable this in the future, than needing to update every use site to enable this now and again in the future to remove it.

Are there many call sites? I was expecting a single change https://github.com/pytorch/pytorch/blob/main/benchmarks/dynamo/common.py#L1500 which would skip the targeted decompositions during torchbench, only

BowenBao commented:
@thiagocrepaldi I see what you meant. I think we should rather apply this by default on exporter. Ideally there shouldn't be difference in the end model people get, between this method and the eventual fix. So this is implementation details.

BowenBao commented:
@pytorchbot merge

BowenBao commented:
@pytorchbot merge
